pkg/common/cluster_actions.go

```diff
@@ -535,6 +555,133 @@ func (i *ClusterActionRunner) configureBrowserRedirector(provider, flow string,
 	return nil
 }
 
+func (i *ClusterActionRunner) configureRealmRequiredActions(obj *v1alpha1.KeycloakRealm) error {
+	realmName := obj.Spec.Realm.Realm
+	actionLogger := log.WithValues("Request.Namespace", obj.Namespace, "Request.Name", obj.Name, "Realm.Name", realmName)
+
+	// Fetch realm required actions from Keycloak API
+	actualRealmRequiredActions, err := i.keycloakClient.ListRealmRequiredActions(realmName)
+	if err != nil {
+		return err
+	}
+
+	actualRealmRequiredActionsAliases, actualRealmRequiredActionsMap := prepareActualRealmRequiredActions(actualRealmRequiredActions)
+	desiredRealmRequiredActionsAliases, desiredRealmRequiredActionsMap := prepareDesiredRealmRequiredActions(obj)
+
+	requiredActionsToRemove := difference(actualRealmRequiredActionsAliases, desiredRealmRequiredActionsAliases)
+	requiredActionsToAdd := difference(desiredRealmRequiredActionsAliases, actualRealmRequiredActionsAliases)
+	requiredActionsToCompare := difference(union(actualRealmRequiredActionsAliases, desiredRealmRequiredActionsAliases), append(requiredActionsToAdd, requiredActionsToRemove...))
+
+	actionLogger.Info(fmt.Sprintf("[REALM REQUIRED ACTION] Adding: %v, Comparing: %v", requiredActionsToAdd, requiredActionsToCompare))
+
+	// Do additions
+	for _, alias := range requiredActionsToAdd {
+		actionLogger.Info(fmt.Sprintf("[REALM REQUIRED ACTION REGISTRATION] Registering new realm required action %v", alias))
+		err = i.keycloakClient.RegisterRealmRequiredAction(desiredRealmRequiredActionsMap[alias], realmName)
+		if err != nil {
+			requiredActionJSON, jsonerr := json.Marshal(desiredRealmRequiredActionsMap[alias])
+			if jsonerr != nil {
+				actionLogger.Info(fmt.Sprintf("[REALM REQUIRED ACTION REGISTRATION - ERROR] Registering new realm required action %v, and unable to marshal JSON.", alias))
+				return err
+			}
+			actionLogger.Info(fmt.Sprintf("[REALM ROLE REGISTRATION - ERROR] Creating new realm role %v", requiredActionJSON))
+			return err
+		}
+	}
+
+	// If additions were made, rebuild the actual realm list.
+	if len(requiredActionsToAdd) > 0 {
```

Comment:
This was required for roles, specifically to support composites: The use case where a single change contains a new role AND adds that new role to another role as a composite.

Is there any scenario where the newly created role needs to be referenced again? You pass the rebuilt map into `compareAndUpdateRealmRequiredAction`, but do you need to or use it?

Same question for clientScope implementation.